pkg/controller/namespace/namespace_controller.go

```diff
@@ -54,510 +68,88 @@ func NewNamespaceController(kubeClient clientset.Interface, versions *unversione
-	items, err := expClient.HorizontalPodAutoscalers(ns).List(api.ListOptions{})
 	if err != nil {
+		glog.Infof("Unable to retrieve namespace %v from store: %v", key, err)
+		nm.queue.Add(key)
 		return err
 	}
-	for i := range items.Items {
-		err := expClient.HorizontalPodAutoscalers(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
-	}
-	return nil
-}
-
-func deleteDaemonSets(expClient extensions_unversioned.ExtensionsInterface, ns string) error {
-	items, err := expClient.DaemonSets(ns).List(api.ListOptions{})
-	if err != nil {
-		return err
-	}
-	for i := range items.Items {
-		err := expClient.DaemonSets(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
-	}
-	return nil
+	namespace := obj.(*api.Namespace)
+	return syncNamespace(nm.kubeClient, nm.versions, namespace)
 }
 
-func deleteJobs(expClient extensions_unversioned.ExtensionsInterface, ns string) error {
-	items, err := expClient.Jobs(ns).List(api.ListOptions{})
-	if err != nil {
-		return err
-	}
-	for i := range items.Items {
-		err := expClient.Jobs(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
-	}
-	return nil
-}
-
-func deleteDeployments(expClient extensions_unversioned.ExtensionsInterface, ns string) error {
-	items, err := expClient.Deployments(ns).List(api.ListOptions{})
-	if err != nil {
-		return err
-	}
-	for i := range items.Items {
-		err := expClient.Deployments(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
-	}
-	return nil
-}
-
-func deleteIngress(expClient extensions_unversioned.ExtensionsInterface, ns string) error {
-	items, err := expClient.Ingresses(ns).List(api.ListOptions{})
-	if err != nil {
-		return err
-	}
-	for i := range items.Items {
-		err := expClient.Ingresses(ns).Delete(items.Items[i].Name, nil)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
-	}
-	return nil
-}
-
-// TODO: this is duplicated logic.  Move it somewhere central?
-func containsVersion(versions *unversioned.APIVersions, version string) bool {
-	for ix := range versions.Versions {
-		if versions.Versions[ix] == version {
-			return true
-		}
-	}
-	return false
-}
-
-// TODO: this is duplicated logic.  Move it somewhere central?
-func containsResource(resources *unversioned.APIResourceList, resourceName string) bool {
-	if resources == nil {
-		return false
-	}
-	for ix := range resources.APIResources {
-		resource := resources.APIResources[ix]
-		if resource.Name == resourceName {
-			return true
-		}
+// Run starts observing the system with the specified number of workers.
+func (nm *NamespaceController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go nm.controller.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(nm.worker, time.Second, stopCh)
```

Comment:
IIUC, each namespace can be in the queue at most once (it can be readded back only after it's already removed from queue right?) And this is the mechanism to enforce that guarantees that there will not be two workers processing the same namesapce at the same time.
Am I right? If so, can you please add a comment about it?